aten/src/ATen/native/mkldnn/xpu/detail/Deconv.h

```diff
@@ -0,0 +1,417 @@
+
+  // construct primitive attr
+  dnnl::primitive_attr pattr;
+  dnnl::post_ops po = attr.extract_post_ops(dst);
+  pattr.set_post_ops(po);
+  #if ONEDNN_SUPPORT_DETERMINISTIC
+    if(at::globalContext().deterministicAlgorithms())
+        pattr.set_deterministic(true);
+  #endif
+
+  pattr.set_scratchpad_mode(dnnl::scratchpad_mode::user);
+
+  auto deconv_fwd_pd = dnnl::deconvolution_forward::primitive_desc(
+      engine,
+      dnnl::prop_kind::forward,
+      dnnl::algorithm::deconvolution_direct,
+      src_md,
+      wgh_md,
+      bia_md,
+      dst_md,
+      _stride,
+      _dilation,
+      _padding,
+      _padding,
+      pattr);
+
+  dnnl::memory src_m, wgh_m, dst_m, bia_m;
+  at::Tensor src_blocked, wgh_blocked, dst_blocked = dst;
+
+  src_m = make_onednn_memory(src_md, engine, src.data_ptr());
+  wgh_m = make_onednn_memory(wgh_md, engine, wgh.data_ptr());
+  dst_m = make_onednn_memory(dst_md, engine, dst.data_ptr());
+
+  std::unordered_map<int, dnnl::memory> args;
+  args.insert({DNNL_ARG_SRC, src_m});
+  args.insert({DNNL_ARG_WEIGHTS, wgh_m});
+  args.insert({DNNL_ARG_DST, dst_m});
+
+  if (bia.defined()) {
+    auto bia_m = make_onednn_memory(bia_md, engine, bia.data_ptr());
+    args.insert({DNNL_ARG_BIAS, bia_m});
+  }
+  if (attr.with_binary())
+    attr.construct_post_binary(deconv_fwd_pd, args);
+
+  size_t scratchpad_size = deconv_fwd_pd.scratchpad_desc().get_size();
+  at::Tensor scratchpad_tensor = at::empty(
+      {static_cast<int64_t>(scratchpad_size)}, src.options().dtype(at::kByte), c10::nullopt);
+  auto scratchpad_m = make_onednn_memory(
+      deconv_fwd_pd.scratchpad_desc(), engine, scratchpad_tensor.data_ptr());
+  args.insert({DNNL_ARG_SCRATCHPAD, scratchpad_m});
+
+  auto deconv_fwd = dnnl::deconvolution_forward(deconv_fwd_pd);
+  XPU_ONEDNN_EXEC(deconv_fwd, stream, args);
+
+}
+
+static void deconvolution_backward_data(
+    at::Tensor& diff_src,
+    const at::Tensor& diff_dst,
+    const at::Tensor& weight,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation,
+    int64_t groups,
+    bool bias_defined) {
+  auto engine =
+      GpuEngineManager::Instance().get_engine({c10::kXPU, c10::xpu::current_device()});
+  auto stream = GpuStreamManager::Instance().get_stream();
+
+  bool is_channels_last_suggested =
+      use_channels_last_for_conv(diff_dst, weight, /*is_transposed=*/true);
+  // create memory desc
+  dnnl::memory::desc src_md, wgh_md, dst_md;
+  std::tie(src_md, wgh_md, dst_md) =
+      deconv_get_plain_md(
+          diff_src, weight, diff_dst, groups, is_channels_last_suggested);
+
+  dnnl::memory::format_tag bia_fmt = dnnl::memory::format_tag::x;
+  auto bia_md = bias_defined
+      ? dnnl::memory::desc({diff_dst.size(1)}, wgh_md.get_data_type(), bia_fmt)
+      : dnnl::memory::desc();
+
+  // create fwd primitive desc hint
+  dnnl::primitive_attr pattr;
+  pattr.set_scratchpad_mode(dnnl::scratchpad_mode::user);
+  #if ONEDNN_SUPPORT_DETERMINISTIC
```

Comment:
why do we need the build macro to control determinism? Shouldn't it be a runtime check?

Reply:
hi, @jgong5, deterministic is valid after oneDNN v3.4, which is newer than the oneDNN version in ideep. The marco here is a compilation check for the `version >= 3.4`.   If oneDNN version requirement is not met, code in the marco would fail during compilation.